pkg/proc/i386_arch.go

```diff
@@ -0,0 +1,422 @@
+		// - cfa is bp + i.PtrSize()*2
+		// - bp is [bp] (i.e. [cfa-i.PtrSize()*2])
+		// - sp is cfa
+
+		// When the signal handler runs it will move the execution to the signal
+		// handling stack (installed using the sigaltstack system call).
+		// This isn't i proper stack switch: the pointer to g in TLS will still
+		// refer to whatever g was executing on that thread before the signal was
+		// received.
+		// Since go did not execute i stack switch the previous value of sp, pc
+		// and bp is not saved inside g.sched, as it normally would.
+		// The only way to recover is to either read sp/pc from the signal context
+		// parameter (the ucontext_t* parameter) or to unconditionally follow the
+		// frame pointer when we get to runtime.sigreturn (which is what we do
+		// here).
+
+		return &frame.FrameContext{
+			RetAddrReg: i386DwarfIPRegNum,
+			Regs: map[uint64]frame.DWRule{
+				i386DwarfIPRegNum: frame.DWRule{
+					Rule:   frame.RuleOffset,
+					Offset: int64(-i.PtrSize()),
+				},
+				i386DwarfBPRegNum: frame.DWRule{
+					Rule:   frame.RuleOffset,
+					Offset: int64(-2 * i.PtrSize()),
+				},
+				i386DwarfSPRegNum: frame.DWRule{
+					Rule:   frame.RuleValOffset,
+					Offset: 0,
+				},
+			},
+			CFA: frame.DWRule{
+				Rule:   frame.RuleCFA,
+				Reg:    i386DwarfBPRegNum,
+				Offset: int64(2 * i.PtrSize()),
+			},
+		}
+	}
+
+	if i.crosscall2fn == nil {
+		i.crosscall2fn = bi.LookupFunc["crosscall2"]
+	}
+
+	// TODO, also some problem about cgo of crosscall2 on 32 bit, need to check.
+	if i.crosscall2fn != nil && pc >= i.crosscall2fn.Entry && pc < i.crosscall2fn.End {
+		rule := fctxt.CFA
+		if rule.Offset == crosscall2SPOffsetBad {
+			switch i.goos {
+			case "windows":
+				rule.Offset += crosscall2SPOffsetWindows
+			default:
+				rule.Offset += crosscall2SPOffsetNonWindows
+			}
+		}
+		fctxt.CFA = rule
+	}
+
+	// We assume that EBP is the frame pointer and we want to keep it updated,
+	// so that we can use it to unwind the stack even when we encounter frames
+	// without descriptor entries.
+	// If there isn't i rule already we emit one.
+	if fctxt.Regs[i386DwarfBPRegNum].Rule == frame.RuleUndefined {
+		fctxt.Regs[i386DwarfBPRegNum] = frame.DWRule{
+			Rule:   frame.RuleFramePointer,
+			Reg:    i386DwarfBPRegNum,
+			Offset: 0,
+		}
+	}
+
+	return fctxt
+}
+
+// cgocallSPOffsetSaveSlot is the offset from systemstack.SP where
+// (goroutine.SP - StackHi) is saved in runtime.asmcgocall after the stack
+// switch happens.
+const i386cgocallSPOffsetSaveSlot = 0x28
+
+// SwitchStack will use the current frame to determine if it's time to
+// switch between the system stack and the goroutine stack or vice versa.
+// Sets it.atend when the top of the stack is reached.
+func (i *I386) SwitchStack(it *stackIterator, _ *op.DwarfRegisters) bool {
+	if it.frame.Current.Fn == nil {
+		return false
+	}
+	switch it.frame.Current.Fn.Name {
+	case "runtime.asmcgocall":
+		if it.top || !it.systemstack {
```

Comment:
From what I can tell from the tests, cgo stacktraces don't work. Therefore I would prefer that SwitchStack didn't try to do things it can't do and simply remove the cases for runtime.asmcgocall and runtime.cgocallback_gofunc for now.